### PR TITLE
feat: [CI-5554]: block malicious patterens in ci command

### PR DESCRIPTION
**Pull request by HenAmar**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37524)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_Failed_** 
 


            
 #### 6 Issues 
* Vulnerability 1 Vulnerabilities 
 * Code issue 5 Code issues 

 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37524)

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java [hunk 1]
```diff
@@ -133,6 +135,15 @@ public void validateResources(Ambiance ambiance, StepElementParameters stepEleme
     if (isNotEmpty(connectorsEntityDetails)) {
       pipelineRbacHelper.checkRuntimePermissions(ambiance, connectorsEntityDetails, true);
     }
+
+    verifySteps(initializeStepInfo);
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java:139]: sanitizeExecution ?

Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java:139]: call it if infra is k8 hosted

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIExecutionValidationServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class CIExecutionValidationServiceImpl implements CIExecutionValidationService {
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIExecutionValidationServiceImpl.java:25]: should be singleton

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIExecutionValidationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  Set<String> maliciousMiningPatterns;
+  static String fileName = "suspiciousMiningPatterns.txt";
+
+  @Inject
+  public CIExecutionValidationServiceImpl() {
+    maliciousMiningPatterns = new HashSet<>();
+    ClassLoader classLoader = getClass().getClassLoader();
+    try {
+      String list =
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIExecutionValidationServiceImpl.java:34]: it should be global variable and static final

**Review by harshjain12 — CHANGES_REQUESTED** [review]

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java (hunk 1 above)
Comment (HenAmar) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java:139]: the validation for infra happens in the function

File: 332-ci-manager/service/src/main/resources/suspiciousMiningPatterns.txt [hunk 4]
```diff
@@ -0,0 +1,394 @@
+mining
+pool
```
Comment (shubham149) [inline 332-ci-manager/service/src/main/resources/suspiciousMiningPatterns.txt:2]: `pool` keyword may fail valid pipelines as well

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java [hunk 5]
```diff
@@ -133,6 +135,15 @@ public void validateResources(Ambiance ambiance, StepElementParameters stepEleme
     if (isNotEmpty(connectorsEntityDetails)) {
       pipelineRbacHelper.checkRuntimePermissions(ambiance, connectorsEntityDetails, true);
     }
+
+    sanitizeExecution(initializeStepInfo);
+  }
+
+  private void sanitizeExecution(InitializeStepInfo initializeStepInfo) {
+    List<ExecutionWrapperConfig> steps = initializeStepInfo.getExecutionElementConfig().getSteps();
+    if (initializeStepInfo.getInfrastructure().getType() == Infrastructure.Type.KUBERNETES_HOSTED) {
```
Comment (shubham149) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java:144]: can you add hosted_vms as well

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIExecutionValidationServiceImpl.java (hunk 2 above)
Comment (HenAmar) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIExecutionValidationServiceImpl.java]: singleton by this bind(CIExecutionValidationService.class).to(CIExecutionValidationServiceImpl.class).in(Singleton.class);

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java (hunk 5 above)
Comment (HenAmar) [reply 332-ci-manager/service/src/main/java/io/harness/ci/execution/states/InitializeTaskStep.java]: done

File: 332-ci-manager/service/src/main/resources/suspiciousMiningPatterns.txt [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+aeon.pool.minergate.com
+arcticrise.com
+asicshack
+at01.supportxmr.com
```
[the remark this answers is not in the document]
Comment (HenAmar) [reply 332-ci-manager/service/src/main/resources/suspiciousMiningPatterns.txt]: done

File: 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIYAMLSanitizationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+  }
+
+  @Override
+  public boolean validate(List<ExecutionWrapperConfig> wrapper) {
+    for (ExecutionWrapperConfig executionWrapper : wrapper) {
+      if (executionWrapper.getStep() == null || executionWrapper.getStep().isNull()) {
+        continue;
+      }
+      StepElementConfig stepElementConfig = IntegrationStageUtils.getStepElementConfig(executionWrapper);
```
Comment (harshjain12) [inline 332-ci-manager/service/src/main/java/io/harness/ci/execution/validation/CIYAMLSanitizationServiceImpl.java:53]: We should not use StepElementConfig

**Review by harshjain12 — APPROVED** [review]
> Reviewed 3 of 5 files at r2, 2 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37524)* status: 2. unresolved discussions. Waiting on team(s): sto.

**Review by shubham149 — APPROVED** [review]
> Reviewed 3 of 5 files at r2, 2 of 3 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37524)* status: Waiting on team(s): sto.

**Review by shubham149 — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37524#-:-NC0N0Y6Bg3OxDk9xEsp:bnfp4nl)* status: Waiting on team(s): sto.

**Pull request merged**